comid-triples.cddl

```diff
@@ -0,0 +1,172 @@
+concise-mid-tag = {
+  ; TODO(tho) language
+  ; TODO(tho) tag-id
+  ; TODO(tho) module-name
+  ; TODO(tho) entity
+  ; TODO(tho) linked-tags
+  comid.triples => triples-map
+  ; TODO(tho) extension
+}
+
+triples-map = non-empty<{
+  ? comid.reference-triples => one-or-more<reference-triple-record>
+  ? comid.endorsed-triples => one-or-more<endorsed-triple-record>
+  ? comid.attest-key-triples => one-or-more<attest-key-triple-record>
+  ? comid.identity-triples => one-or-more<identity-triple-record>
+  * $$triples-map-extension
+}>
+
+; REFERENCE-VALUE triple
+; "target environment $TE" "has reference measurements" "$RV"
+; NOTE(tho): can an attesting evironement have reference values?
+reference-triple-record = [
+  environment-map ; target environment
+  one-or-more<measurement-map> ; reference measurements
+]
+
+; ENDORSED-VALUE triple
+; "environment $E" "has endorsed measurements" "$EV"
+endorsed-triple-record = [
+  environment-map ; (target or attesting) environment
+  one-or-more<measurement-map> ; endorsed measurements
+]
+
+; ATTESTATION-VERIFICATION-KEY triple
+; "attesting environment $AE" "signs Evidence that can be verified with key" "$K"
+attest-key-triple-record = [
+  environment-map ; attesting environment
+  one-or-more<verification-key-map> ; attestation verification key(s)
+]
+
+; DEVICE-IDENTITY triple
+; "device $D" "is identified by key" "$K"
+identity-triple-record = [
+  environment-map ; device identifier (instance or class)
+  one-or-more<verification-key-map> ; DevID, or semantically equivalent
+]
+
+verification-key-map = {
+  ; Verification key (typically a public key)
+  comid.key => COSE_Key
+  ; Optional X.509 certificate chain corresponding to the public key
+  ; in comid.key, encoded as an array of one or more base64-encoded
+  ; DER PKIX certificates.  The certificate containing the public key
+  ; in comid.key MUST be the first certificate.  This MAY be followed
+  ; by additional ; certificates, with each subsequent certificate
+  ; being the one used to certify the previous one.
+  ? comid.keychain => [ + tstr ]
+}
+
+environment-map = non-empty<{
+  ? comid.class => class-map
+  ? comid.instance => $instance-id-type-choice
+}>
+
+class-map = {
+  comid.class-id => $class-id-type-choice
```

Comment:
Given class-id is no longer optional and there are shipping products that don't include class-id in evidence. The matching algorithm seems to be impacted:
•	For each claim CX in CREF (where claim refers to a class-map structure)
o	Examine CEV for evidence with the same key name as CX and assign it to CY.
	IF CX or CY has a name with a ‘layer’ value that is greater than the ‘layer’ value for any ACCEPTED CLAIMS having the same ‘vendor’, ‘class-id’, ‘model’ AND has different measurements THEN  fail claims verification.

If class-id is omitted from evidence, it cannot be omitted from ref values given current CDDL. This means the matching algorithm breaks. 

If the normative for which fields to include is in a spec rather than in the data definition, then we might say something like "class-id SHOULD be included" whereas current semantics are saying it MUST be included. 

I think it is problematic to use tagged-model-class-type because it is not possible to distinguish the 'ns-auth' aka 'vendor' string from the 'model' string.

Reply:
> If class-id is omitted from evidence, it cannot be omitted from ref values given current CDDL. This means the matching algorithm breaks.

We could have:
```
class-map = non-empty<{
  ? comid.class-id => $class-id-type-choice
  ...
}>
```
and have normative text that you suggest above about -- i.e., "class-map SHOULD include class-id"

Would that work for you?

Reply:
> I think it is problematic to use tagged-model-class-type because it is not possible to distinguish the 'ns-auth' aka 'vendor' string from the 'model' string.

I am slightly confused.  `tagged-model-class-type` is
```
[
  tstr, ; ns-auth == vendor
  tstr ; model
]
```
which allows precisely distinguishing vendor from model.

So I guess you are referring to something else?  Maybe the `comid.vendor` and `comid.model` in the `class-map`?

Reply:
maybe I'm reading the <x> macro wrong. It looked to me like the two strings were concatenated.

Reply:
> > If class-id is omitted from evidence, it cannot be omitted from ref values given current CDDL. This means the matching algorithm breaks.
> 
> We could have:
> 
> ```
> class-map = non-empty<{
>   ? comid.class-id => $class-id-type-choice
>   ...
> }>
> ```
> 
> and have normative text that you suggest above about -- i.e., "class-map SHOULD include class-id"
> 
> Would that work for you?

Yes

Reply:
> maybe I'm reading the macro wrong. It looked to me like the two strings were concatenated.

no, they are encoded as separate items:
```
[ 'vendor X' , 'model Y' ]
```
would be encoded as:
```
82                     # array(2)
   48                  # bytes(8)
      76656E646F722058 # "vendor X"
   47                  # bytes(7)
      6D6F64656C2059   # "model Y"
```

Reply:
If dicetcbinfo encoded 'vendor' and 'model' but not 'type' and CoMID encoded ['vendor' , 'model'] but not 'vendor' and 'model'. The current matching algorithm would not match this (but could).

Reply:
as discussed, we drop the name-spaced types (at least for now)